lib/GADS.pm

```diff
@@ -2753,6 +2753,235 @@ prefix '/:layout_name' => sub {
+
+            $params->{header_back_url} = "${base_url}table";
+            $params->{breadcrumbs}     = [
+                Crumb( $base_url . "table/", "Tables" ),
+                Crumb( "",                   "Table: " . $layout->name )
+            ];
+
+            my $layout_id = $layout->{instance_id};
+
+            my $reports = $layout->reports;
+
+            $params->{reports} = $reports;
+
+            template 'reports/view' => $params;
+        };
+
+        #add a report
+        any [ 'get', 'post' ] => '/add' => require_login sub {
+            if ( body_parameters && body_parameters->{submit} ) {
+                my $layout             = var('layout') or pass;
+                my $report_description = params->{'report_description'};
+                my $report_name        = params->{'report_name'};
+                my $checkbox_fields    = params->{'checkboxes'};
+                my $instance           = $layout->{instance_id}
+                  if $layout && $layout->{instance_id};
+
+                my $user = logged_in_user;
+
+                my $report = schema->resultset('Report')->create_report(
+                    {
+                        schema      => schema,
+                        user        => $user,
+                        name        => $report_name,
+                        description => $report_description,
+                        instance_id => $instance,
+                        createdby   => $user,
+                        layouts     => $checkbox_fields
+                    }
+                );
+
+                my $lo = param 'layout_name';
+                return forwardHome( { success => "Report created" },
+                    "$lo/report" );
+            }
+
+            my $user   = logged_in_user;
+            my $layout = var('layout') or pass;
+
+            if ( app->has_hook('plugin.linkspace.data_before_request') ) {
+                app->execute_hook( 'plugin.linkspace.data_before_request',
+                    user => $user );
+            }
+
+            my %params = (
+                user   => $user,
+                layout => $layout,
+                schema => schema,
+            );
+
+            my $records = [ $layout->all( user_can_read => 1 ) ];
+
+            my $base_url = request->base;
+
+            my $params;
+
+            $params->{header_type} = 'table_tabs';
+
+            $params->{layout_obj} = $layout;
+            $params->{layout}     = $layout;
+
+            $params->{header_back_url} = "${base_url}table";
+            $params->{breadcrumbs}     = [
+                Crumb( $base_url . "table/", "Tables" ),
+                Crumb( "",                   "Table: " . $layout->name )
+            ];
+
+            $params->{viewtype} = 'add';
+            $params->{fields}   = $records;
+
+            template 'reports/edit' => $params;
+        };
+
+        #Edit a report (by :id)
+        any [ 'get', 'post' ] => '/edit:id' => require_login sub {
```

Comment:
I think most comments in the previous route also apply to this route. In the main I think it can be collapsed down to be a lot shorter with some of the code removed.